### DISCUSSION TITLE
Cleanup of temporary files after exports have been processed/uploaded

### DISCUSSION THREAD

**Discussion by soulchild** [outline/outline#5450] [Ideas]
My understanding is that before any exports are uploaded to S3, the ZIP archives are written to the `/tmp` directory.

Unfortunately, the temporary files are never explicitly cleaned up which might potentially lead to disk space issues. IMHO leaving this to the OS is not optimal, because tmp cleanup only happens when rebooting the machine/VM - or restarting the Docker container in our case.

I tracked down the export functionality to this [code path](https://github.com/outline/outline/blob/9ef375d83cfa13a1c2addeed740e024bdd84c203/server/queues/tasks/ExportTask.ts#L50). It should be easy to delete the temporary file returned by `this.export()` once the S3 upload succeeds. Would you be interested in a PR to address this? Cheers!

**Comment by tommoor** [comment]
Sure, yep. I’d put the cleanup in a finally block to catch possible failure scenarios

**Comment by soulchild** [comment]
https://github.com/outline/outline/pull/5456

**Comment by soulchild** [comment]
PR #5456 has been merged

**Comment by soulchild** [comment]
@tommoor We are seeing a lot of open file handles referencing the `/tmp` directory since this has landed. The temporary files are indeed getting deleted, but they're still occupying space (until the outline container is restarted).

I suppose the file handles of the `unlink` operation stay open or something like that. Have you seen this behavior before? Maybe it'd help if we awaited the `unlink` operation instead of just [fire-and-forgetting](https://github.com/outline/outline/blob/7e17e82ac84e60a8b62dd126eb8b70336f4a9c4f/server/queues/tasks/ExportTask.ts#L132) it?

**Reply by tommoor** [reply]
`await` ing the promise wouldn't make a difference to this behavior, could be worth truncating the file before unlinking. But if file handles are handing around at the OS level for unlinked files that sounds more like a nodejs bug? 😬

**Reply by soulchild** [reply]
Yeah, I'm also pretty sure `await`ing the promise won't make a difference. Maybe the exported file needs to be explicitly [closed](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#filehandleclose)? I'll try to find out whether this helps.

**Comment by Coding-Kiwi** [comment]
Hi, this is still an issue.

My server drive was at 90% and I tried to find out what was happening. Finally running `sudo lsof +L1` revealed all the files deleted by the process but still kept open

![image](https://github.com/user-attachments/assets/6019d7f6-ff97-4f1d-9e97-2be944b52897)

which I could trace back to outline

```sh
cat /proc/3164060/cgroup

#output is something like 0::/system.slice/docker-ae5f85453478c803c63b315543a79713221607e2fb971586f15c72cc84516124.scope

docker ps --filter "id=ae5f85453478c803c63b315543a79713221607e2fb971586f15c72cc84516124"

# --> it is the outline container
```
